test/extensions/filters/http/common/jwks_fetcher_test.cc

```diff
@@ -0,0 +1,209 @@
+      "alg": "RS256",
+      "use": "sig",
+      "kid": "b3319a147514df7ee5e4bcdee51350cc890cc89e",
+      "n": "up97uqrF9MWOPaPkwSaBeuAPLOr9FKcaWGdVEGzQ4f3Zq5WKVZowx9TCBxmImNJ1qmUi13pB8otwM_l5lfY1AFBMxVbQCUXntLovhDaiSvYp4wGDjFzQiYA-pUq8h6MUZBnhleYrkU7XlCBwNVyN8qNMkpLA7KFZYz-486GnV2NIJJx_4BGa3HdKwQGxi2tjuQsQvao5W4xmSVaaEWopBwMy2QmlhSFQuPUpTaywTqUcUq_6SfAHhZ4IDa_FxEd2c2z8gFGtfst9cY3lRYf-c_ZdboY3mqN9Su3-j3z5r2SHWlhB_LNAjyWlBGsvbGPlTqDziYQwZN4aGsqVKQb9Vw",
+      "e": "AQAB"
+    }
+  ]
+}
+)";
+
+const std::string JwksUri = R"(
+uri: https://pubkey_server/pubkey_path
+cluster: pubkey_cluster
+timeout:
+  seconds: 5
+)";
+
+class JwksFetcherTest : public ::testing::Test {
+public:
+  void SetUp() { MessageUtil::loadFromYaml(JwksUri, uri_); }
+  HttpUri uri_;
+  testing::NiceMock<Server::Configuration::MockFactoryContext> mock_factory_ctx_;
+};
+
+// A mock HTTP upstream with response body.
+class MockUpstream {
+public:
+  MockUpstream(Upstream::MockClusterManager& mock_cm, const std::string& status,
+               const std::string& response_body)
+      : request_(&mock_cm.async_client_), status_(status), response_body_(response_body) {
+    ON_CALL(mock_cm.async_client_, send_(testing::_, testing::_, testing::_))
+        .WillByDefault(testing::Invoke([this](Http::MessagePtr&, Http::AsyncClient::Callbacks& cb,
+                                              const absl::optional<std::chrono::milliseconds>&)
+                                           -> Http::AsyncClient::Request* {
+          Http::MessagePtr response_message(new Http::ResponseMessageImpl(
+              Http::HeaderMapPtr{new Http::TestHeaderMapImpl{{":status", status_}}}));
+          if (response_body_.length()) {
+            response_message->body().reset(new Buffer::OwnedImpl(response_body_));
+          } else {
+            response_message->body().reset(nullptr);
+          }
+          cb.onSuccess(std::move(response_message));
+          return &request_;
+        }));
+  }
+
+  MockUpstream(Upstream::MockClusterManager& mock_cm, Http::AsyncClient::FailureReason reason)
+      : request_(&mock_cm.async_client_) {
+    ON_CALL(mock_cm.async_client_, send_(testing::_, testing::_, testing::_))
+        .WillByDefault(testing::Invoke(
+            [this, reason](
+                Http::MessagePtr&, Http::AsyncClient::Callbacks& cb,
+                const absl::optional<std::chrono::milliseconds>&) -> Http::AsyncClient::Request* {
+              cb.onFailure(reason);
+              return &request_;
+            }));
+  }
+
+  MockUpstream(Upstream::MockClusterManager& mock_cm, Http::MockAsyncClientRequest* request)
+      : request_(&mock_cm.async_client_) {
+    ON_CALL(mock_cm.async_client_, send_(testing::_, testing::_, testing::_))
+        .WillByDefault(testing::Return(request));
+  }
+
+private:
+  Http::MockAsyncClientRequest request_;
+  std::string status_;
+  std::string response_body_;
+};
+
+class MockJwksReceiver : public JwksFetcher::JwksReceiver {
```

Comment:
Please move this mock to mocks.h  so it can be shared by other tests.

Reply:
Fixed in 8038924